lib/Sema/TypoCorrection.h

```diff
@@ -0,0 +1,98 @@
+//===--- TypoCorrection.h - Typo correction ---------------------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+//  This file defines the interface for doing typo correction.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_SEMA_TYPOCORRECTION_H
+#define SWIFT_SEMA_TYPOCORRECTION_H
+
+#include "swift/Basic/LLVM.h"
+#include "swift/AST/Identifier.h"
+#include "swift/AST/DeclNameLoc.h"
+#include "llvm/ADT/Optional.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/STLExtras.h"
+
+namespace swift {
+
+class LookupResult;
+class TypeChecker;
+
+/// A typo correction to perform.
+class TypoCorrection {
+public:
+  DeclName WrittenName;
+  DeclNameLoc Loc;
+  DeclName CorrectedName;
+
+  TypoCorrection(DeclName writtenName, DeclNameLoc writtenLoc,
+                 DeclName correctedName)
+    : WrittenName(writtenName), Loc(writtenLoc), CorrectedName(correctedName) {}
+
+  void addFixits(InFlightDiagnostic &diagnostic) const;
+};
+
+/// A collection of typo-correction candidates.
+class TypoCorrectionResults {
+public:
+  TypeChecker &TC;
+  DeclName WrittenName;
+  DeclNameLoc Loc;
+
+  SmallVector<ValueDecl *, 4> Candidates;
+
+  TypoCorrectionResults(TypeChecker &tc, DeclName writtenName, DeclNameLoc loc)
+    : TC(tc), WrittenName(writtenName), Loc(loc) {}
+
+  /// Try to claim a unique correction from this collection that's simple
+  /// enough to include "inline" in the primary diagnostic.
+  ///
+  /// The expected pattern is that CorrectedName will be added to the
+  /// diagnostic (as in "did you mean <<CorrectedName>>"), and then addFixits
+  /// will be called on the still-in-flight diagnostic.
+  ///
+  /// If this returns a correction, it clears the candidate set so that
+  /// callers can freely call noteAllCandidates without remembering that
+  /// they did this.
+  Optional<TypoCorrection> claimUniqueCorrection();
+
+  /// Emit a note for every candidate in the set.
+  void noteAllCandidates() const;
+
+  /// Add all the candidates to a lookup set.
+  void addAllCandidatesToLookup(LookupResult &lookup) const;
+
+  /// Look for a single candidate in this set that matches the given predicate.
+  template <class Fn>
+  ValueDecl *getUniqueCandidateMatching(Fn &&predicate) {
```

Comment:
The API here is a bit inconsistent. `claimUniqueCorrection()` only produces the name, so the client needs to perform lookup again with the corrected name. `getUniqueCandidateMatching()` doesn't de-duplicate things with the same name, but does return the declaration... which skips the extra lookup in the client (and the validation that goes with performing that lookup directly). It seems to me like the "matching" variant should work the same as `claimUniqueCorrection()` (de-duplicate and only return the corrected name), with the added predicate check.

Reply:
I've revised this a bit.  The candidates are now left in the TypoCorrectionResults, and we just change how we note them.  This means that recovery can proceed by just taking the candidates out of the results set.

A TypoCorrection should not carry a candidate because we might have a unique correction that corresponds to multiple candidates.  This should also be clearer now, I hope.